qiskit/primitives/containers/bindings_array.py

```diff
@@ -0,0 +1,357 @@
+    arrays attached to parameters, ``{params0: kwvals0, ...}``. Crucially, the last dimension of
+    each array indexes one or more parameters. For example, if the last dimension of ``vals1`` is
+    25, then it represents an array of possible binding values for 25 distinct parameters, where its
+    leading shape is the array :attr:`~.shape` of its binding array. This implies a degeneracy of the
+    storage format: ``[vals, vals1[..., :10], vals1[..., 10:], ...]`` is exactly equivalent to
+    ``[vals0, vals1, ...]`` in the bindings it specifies. This complication has been included to
+    satisfy two competing constraints:
+
+        * Arrays with different dtypes cannot be concatenated into a single array, so that multiple
+          arrays are required for generality.
+        * It is extremely convenient to put everything into a small number of big arrays, when
+          possible.
+
+    .. code-block:: python
+
+        # 0-d array (i.e. only one binding)
+        BindingsArray([1, 2, 3], {"a": 4, ("b", "c"): [5, 6]})
+
+        # single array, last index is parameters
+        BindingsArray(np.empty((10, 10, 100)))
+
+        # multiple arrays, where each last index is parameters. notice that it's smart enough to
+        # figure out that a missing last dimension corresponds to a single parameter.
+        BindingsArray(
+            [np.empty((10, 10, 100)), np.empty((10, 10)), np.empty((10, 10, 20), dtype=complex)],
+            {("c", "a"): np.empty((10, 10, 2)), "b": np.empty((10, 10))}
+        )
+    """
+    __slots__ = ("_vals", "_kwvals")
+
+    def __init__(
+        self,
+        vals: ArrayLike | Iterable[ArrayLike] | None = None,
+        kwvals: Mapping[ParameterLike, Iterable[ParameterValueType]] | ArrayLike | None = None,
+        shape: ShapeInput | None = None,
+    ):
+        """
+        Initialize a ``BindingsArray``. It can take parameter vectors and dictionaries.
+
+        The ``shape`` argument does not need to be provided whenever it can unambiguously
+        be inferred from the provided arrays. Ambiguity arises because an array provided to the
+        constructor might represent values for either a single parameter, with an implicit missing
+        last dimension of size ``1``, or for many parameters, where the size of the last dimension
+        is the number of parameters it is providing values to. This ambiguity can be broken in the
+        following common ways:
+
+            * Only a single array is provided to ``vals``, and no arrays to ``kwvals``, in which case
+              it is assumed that the last dimension is over many parameters.
+            * Multiple arrays are given whose shapes differ only in the last dimension size.
+            * Some array is given in ``kwvals`` where the key contains multiple
+              :class:`~.Parameter` s, whose length the last dimension of the array must therefore match.
```

Comment:
ReST is picky about which characters are allowed to immediately follow a role, but in general, a backslash is allowed.

```suggestion
              :class:`~.Parameter`\s, whose length the last dimension of the array must therefore match.
```